File: Rectangle.Windows/src/Rectangle.Windows.WinUI.Rust/src/services/window_manager.rs
use crate::core::action::{SubsequentExecutionMode, WindowAction};
use crate::core::calculator_factory::CalculatorFactory;
use crate::core::history::WindowHistory;
use crate::core::rect::{WindowRect, WorkArea};
use crate::services::config::ConfigService;
use crate::services::logger::Logger;
use crate::services::win32::Win32WindowService;
use std::collections::HashSet;
use std::sync::{Arc, Mutex};

/// 窗口管理器
pub struct WindowManager {
    config_service: ConfigService,
    win32_service: Win32WindowService,
    calculator_factory: CalculatorFactory,
    window_history: Arc<Mutex<WindowHistory>>,
    ignored_apps: Arc<Mutex<HashSet<String>>>,
}

impl WindowManager {
    /// 创建新的窗口管理器
    pub fn new(config_service: ConfigService) -> Self {
        let calculator_factory = CalculatorFactory::new(Some(&config_service));
        let config = config_service.load();
        let ignored_apps: HashSet<String> = config.ignored_apps.iter().cloned().collect();

        Self {
            config_service,
            win32_service: Win32WindowService::new(),
            calculator_factory,
            window_history: Arc::new(Mutex::new(WindowHistory::new())),
            ignored_apps: Arc::new(Mutex::new(ignored_apps)),
        }
    }

    /// 执行窗口动作
    pub fn execute(&self, action: WindowAction, force_direct_action: bool) -> bool {
        let hwnd = self.win32_service.get_foreground_window_handle();
        if hwnd == 0 {
            Logger::warning("WindowManager", "没有活动窗口");
            return false;
        }

        // 检查窗口是否被忽略
        if self.is_app_ignored(hwnd) {
            Logger::info("WindowManager", "窗口在忽略列表中，跳过");
            return false;
        }

        // 获取当前窗口矩形
        let current_rect = match self.win32_service.get_window_rect(hwnd) {
            Some(rect) => WindowRect::new(rect.0, rect.1, rect.2, rect.3),
            None => {
                Logger::warning("WindowManager", "无法获取窗口矩形");
                return false;
            }
        };

        // 获取工作区
        let work_area = match self.win32_service.get_work_area_from_window(hwnd) {
            Some(wa) => wa,
            None => {
                Logger::warning("WindowManager", "无法获取工作区");
                return false;
            }
        };

        // 获取配置
        let config = self.config_service.load();
        let gap = config.gap_size;

        // 处理特殊动作
        match action {
            WindowAction::Undo => return self.undo(hwnd),
            WindowAction::Redo => return self.redo(hwnd),
            WindowAction::Restore => return self.restore(hwnd),
            WindowAction::Maximize => {
                self.win32_service.show_window_maximize(hwnd);
                return true;
            }
            _ => {}
        }

        // 保存恢复点
        {
            let mut history = self.window_history.lock().unwrap();
            history.save_restore_rect_if_not_exists(hwnd, current_rect.x, current_rect.y, current_rect.width, current_rect.height);
        }

        // 获取计算器
        let calculator = match self.calculator_factory.get_calculator(action) {
            Some(calc) => calc,
            None => {
                Logger::warning("WindowManager", &format!("未找到动作 {:?} 的计算器", action));
                return false;
            }
        };

        // 计算新位置
        let new_rect = calculator.calculate(&work_area, &current_rect, action, gap);

        // 应用新位置
        let result = self.win32_service.set_window_rect(
            hwnd,
            new_rect.x,
            new_rect.y,
            new_rect.width,
            new_rect.height,
        );

        if result {
            // 记录操作
            let mut history = self.window_history.lock().unwrap();
            history.record_action(hwnd, action, new_rect.x, new_rect.y, new_rect.width, new_rect.height);
            history.mark_as_program_adjusted(hwnd);

            Logger::info("WindowManager", &format!("执行动作 {:?}: ({}, {}, {}, {})",
                action, new_rect.x, new_rect.y, new_rect.width, new_rect.height));
        }

        result
    }

    /// 撤销
    fn undo(&self, hwnd: isize) -> bool {
        let mut history = self.window_history.lock().unwrap();
        if let Some(action) = history.try_get_last_action(hwnd) {
            let (x, y, w, h) = action.rect;
            // 恢复到之前的状态
            let _ = self.win32_service.set_window_rect(hwnd, x, y, w, h);
            history.remove_last_action(hwnd);
            Logger::info("WindowManager", "执行撤销");
            true
        } else {
            Logger::warning("WindowManager", "没有可撤销的操作");
            false
        }
    }

    /// 重做
    fn redo(&self, _hwnd: isize) -> bool {
        // TODO: 实现重做功能
        Logger::info("WindowManager", "重做功能尚未实现");
        false
    }

    /// 恢复
    fn restore(&self, hwnd: isize) -> bool {
        let history = self.window_history.lock().unwrap();
        if let Some((x, y, w, h)) = history.try_get_restore_rect(hwnd) {
            let result = self.win32_service.set_window_rect(hwnd, x, y, w, h);
            Logger::info("WindowManager", "执行恢复");
            result
        } else {
            Logger::warning("WindowManager", "没有可恢复的窗口位置");
            false
        }
    }

    /// 检查应用是否在忽略列表中
    fn is_app_ignored(&self, hwnd: isize) -> bool {
        let process_name = self.win32_service.get_process_name_from_window(hwnd);
        let ignored = self.ignored_apps.lock().unwrap();
        ignored.iter().any(|app| {
            process_name.eq_ignore_ascii_case(app)
                || process_name.eq_ignore_ascii_case(&format!("{}.exe", app))
        })
    }

    /// 刷新忽略列表
    pub fn refresh_ignored_apps(&self) {
        let config = self.config_service.load();
        let mut ignored = self.ignored_apps.lock().unwrap();
        *ignored = config.ignored_apps.iter().cloned().collect();
    }
}

/// 操作历史管理器
pub struct OperationHistoryManager {
    history: Vec<(isize, WindowRect)>,
    current_index: usize,
    max_size: usize,
}

impl OperationHistoryManager {
    /// 创建新的操作历史管理器
    pub fn new(max_size: usize) -> Self {
        Self {
            history: Vec::new(),
            current_index: 0,
            max_size,
        }
    }

    /// 记录操作
    pub fn record(&mut self, hwnd: isize, rect: WindowRect) {
        // 移除当前位置之后的历史
        if self.current_index < self.history.len() {
            self.history.truncate(self.current_index);
        }

        self.history.push((hwnd, rect));
        self.current_index += 1;

        // 限制历史大小
        if self.history.len() > self.max_size {
            self.history.remove(0);
            self.current_index -= 1;
        }
    }

    /// 撤销
    pub fn undo(&mut self) -> Option<(isize, WindowRect)> {
        if self.current_index > 0 {
            self.current_index -= 1;
            self.history.get(self.current_index).copied()
        } else {
            None
        }
    }

    /// 重做
    pub fn redo(&mut self) -> Option<(isize, WindowRect)> {
        if self.current_index < self.history.len() {
            let result = self.history.get(self.current_index).copied();
            self.current_index += 1;
            result
        } else {
            None
        }
    }

    /// 是否可以撤销
    pub fn can_undo(&self) -> bool {
        self.current_index > 0
    }

    /// 是否可以重做
    pub fn can_redo(&self) -> bool {
        self.current_index < self.history.len()
    }
}

File: Rectangle.Windows/src/Rectangle.Windows.WinUI.Cpp/Services/WindowManager.cpp
#include "pch.h"
#include "Services/WindowManager.h"
#include "Services/Logger.h"
#include "Services/Win32WindowService.h"
#include "Services/OperationHistoryManager.h"
#include "Services/LastActiveWindowService.h"
#include "Core/CalculatorFactory.h"
#include "Core/WindowHistory.h"
#include "Services/ConfigService.h"

namespace winrt::Rectangle::Services
{
    WindowManager::WindowManager()
    {
        Logger::Instance().Info(L"WindowManager", L"WindowManager initialized");
    }

    WindowManager::~WindowManager() = default;

    void WindowManager::SetConfigService(ConfigService* configService)
    {
        m_configService = configService;
        ReloadConfig();
    }

    void WindowManager::SetLastActiveWindowService(void* service)
    {
        m_lastActiveService = service;
    }

    void WindowManager::SetOperationHistory(void* history)
    {
        m_operationHistory = history;
    }

    void WindowManager::ReloadConfig()
    {
        if (!m_configService) return;

        auto config = m_configService->Load();
        m_gapSize = config.GapSize;
    }

    bool WindowManager::IsIgnoredApp(const std::wstring& processName) const
    {
        if (!m_configService) return false;

        auto config = m_configService->Load();
        for (const auto& app : config.IgnoredApps)
        {
            if (_wcsicmp(app.c_str(), processName.c_str()) == 0 ||
                _wcsicmp((app + L".exe").c_str(), processName.c_str()) == 0)
            {
                return true;
            }
        }
        return false;
    }

    void WindowManager::Execute(WindowAction action, int64_t targetHwnd, bool forceDirectAction)
    {
        if (action == WindowAction::Restore)
        {
            ExecuteRestore(targetHwnd);
            return;
        }

        if (action == WindowAction::Maximize)
        {
            ExecuteMaximizeToggle(targetHwnd);
            return;
        }

        if (action == WindowAction::NextDisplay)
        {
            ExecuteNextDisplay(targetHwnd);
            return;
        }

        if (action == WindowAction::PreviousDisplay)
        {
            ExecutePreviousDisplay(targetHwnd);
            return;
        }

        if (action == WindowAction::Undo || action == WindowAction::Redo)
        {
            if (action == WindowAction::Undo) ExecuteUndo();
            else ExecuteRedo();
            return;
        }

        int64_t hwnd = targetHwnd ? targetHwnd : GetTargetWindow();
        if (hwnd == 0)
        {
            PlayBeep();
            return;
        }

        Win32WindowService win32;
        auto processName = win32.GetProcessNameFromWindow(hwnd);
        if (processName.empty())
        {
            Logger::Instance().Warning(L"WindowManager", L"无法获取窗口进程名");
            return;
        }

        if (IsIgnoredApp(processName))
        {
            Logger::Instance().Info(L"WindowManager", processName + L" 在忽略列表中，跳过操作");
            return;
        }

        if (win32.IsModalDialog(hwnd))
        {
            Logger::Instance().Info(L"WindowManager", processName + L" 是模态对话框，跳过操作");
            return;
        }

        int32_t x, y, width, height;
        if (!win32.GetWindowRect(hwnd, x, y, width, height))
        {
            return;
        }

        bool isMaximized = win32.IsMaximized(hwnd);
        bool useDirectResizeFromMaximized = false;

        auto isResizeAction = action == WindowAction::Larger || action == WindowAction::Smaller ||
            action == WindowAction::LargerWidth || action == WindowAction::SmallerWidth ||
            action == WindowAction::LargerHeight || action == WindowAction::SmallerHeight;

        if (isResizeAction && isMaximized)
        {
            bool isLarger = action == WindowAction::Larger || action == WindowAction::LargerWidth || action == WindowAction::LargerHeight;
            if (isLarger)
            {
                return;
            }
            useDirectResizeFromMaximized = true;
        }

        auto workArea = GetTargetWorkArea(hwnd);
        Core::WindowRect current(x, y, width, height);

        WindowHistory history;
        bool windowMovedExternally = history.IsWindowMovedExternally(hwnd, x, y, width, height);

        if (windowMovedExternally)
        {
            history.RemoveLastAction(hwnd);
            Logger::Instance().Info(L"WindowManager", L"检测到窗口被用户手动移动: " + processName);
        }

        auto [actualAction, targetDisplayIndex] = forceDirectAction
            ? std::make_pair(action, std::optional<int32_t>{})
            : GetActualAction(hwnd, action, windowMovedExternally);

        if (actualAction != action || targetDisplayIndex.has_value())
        {
            Logger::Instance().Info(L"WindowManager",
                targetDisplayIndex.has_value()
                    ? L"循环尺寸(显示器" + std::to_wstring(targetDisplayIndex.value() + 1) + L"): " + Core::ToString(action) + L" → " + Core::ToString(actualAction)
                    : L"循环尺寸: " + Core::ToString(action) + L" → " + Core::ToString(actualAction));
        }

        CalculatorFactory factory;
        auto calculator = factory.GetCalculator(actualAction);
        if (!calculator)
        {
            Logger::Instance().Warning(L"WindowManager", L"未找到对应的计算器: " + Core::ToString(actualAction));
            return;
        }

        if (targetDisplayIndex.has_value())
        {
            workArea = GetWorkAreaByDisplayIndex(targetDisplayIndex.value());
        }

        if (!history.HasRestoreRect(hwnd) || windowMovedExternally)
        {
            history.SaveRestoreRect(hwnd, x, y, width, height);
            Logger::Instance().Debug(L"WindowManager", L"保存恢复点: (" + std::to_wstring(x) + L", " + std::to_wstring(y) + L", " + std::to_wstring(width) + L", " + std::to_wstring(height) + L")");
        }

        history.MarkAsProgramAdjusted(hwnd);
        m_maximizedWindows.erase(hwnd);

        auto target = calculator->Calculate(workArea, current, actualAction, m_gapSize);
        target = ApplyWindowGap(target, workArea, actualAction);

        if (!win32.IsResizable(hwnd))
        {
            target = HandleFixedSizeWindow(current, target, workArea, actualAction);
        }

        target = ApplyMinimumSize(target);
        target = ClampToWorkArea(target, workArea);

        if (useDirectResizeFromMaximized)
            win32.SetWindowRectFromMaximized(hwnd, target.X, target.Y, target.Width, target.Height);
        else
            win32.SetWindowRect(hwnd, target.X, target.Y, target.Width, target.Height);

        history.RecordAction(hwnd, action, target.X, target.Y, target.Width, target.Height);

        RecordToOperationHistory(action, hwnd, current, target, processName);

        MoveCursorIfEnabled(hwnd, actualAction);

        Logger::Instance().Info(L"WindowManager", Core::ToString(actualAction) + L" 了 " + processName);
    }

    void WindowManager::ExecuteRestore(int64_t targetHwnd)
    {
        int64_t hwnd = targetHwnd ? targetHwnd : GetTargetWindow();
        if (hwnd == 0) return;

        Win32WindowService win32;
        WindowHistory history;

        if (!history.HasRestoreRect(hwnd))
        {
            Logger::Instance().Debug(L"WindowManager", L"窗口没有恢复点");
            return;
        }

        auto restoreRect = history.GetRestoreRect(hwnd);
        win32.SetWindowRect(hwnd, restoreRect.X, restoreRect.Y, restoreRect.Width, restoreRect.Height);
        history.ClearRestoreRect(hwnd);
        m_maximizedWindows.erase(hwnd);

        Logger::Instance().Info(L"WindowManager", L"已恢复窗口");
    }

    void WindowManager::ExecuteMaximizeToggle(int64_t targetHwnd)
    {
        int64_t hwnd = targetHwnd ? targetHwnd : GetTargetWindow();
        if (hwnd == 0) return;

        Win32WindowService win32;

        if (win32.IsMaximized(hwnd))
        {
            ExecuteRestore(hwnd);
        }
        else
        {
            auto workArea = GetTargetWorkArea(hwnd);
            win32.SetWindowRect(hwnd, workArea.Left, workArea.Top, workArea.Width(), workArea.Height());
            m_maximizedWindows.insert(hwnd);
        }
    }

    void WindowManager::ExecuteNextDisplay(int64_t targetHwnd)
    {
        int64_t hwnd = targetHwnd ? targetHwnd : GetTargetWindow();
        if (hwnd == 0) return;

        Win32WindowService win32;
        auto workAreas = win32.GetMonitorWorkAreas();
        if (workAreas.size() <= 1) return;

        auto currentWorkArea = GetTargetWorkArea(hwnd);
        int32_t currentIndex = -1;

        for (size_t i = 0; i < workAreas.size(); ++i)
        {
            if (workAreas[i].Left == currentWorkArea.Left &&
                workAreas[i].Top == currentWorkArea.Top)
            {
                currentIndex = static_cast<int32_t>(i);
                break;
            }
        }

        int32_t nextIndex = (currentIndex + 1) % workAreas.size();
        auto nextWorkArea = workAreas[nextIndex];

        int32_t x, y, width, height;
        if (!win32.GetWindowRect(hwnd, x, y, width, height)) return;

        int32_t newX = nextWorkArea.Left + (x - currentWorkArea.Left);
        int32_t newY = nextWorkArea.Top + (y - currentWorkArea.Top);

        newX = std::max(nextWorkArea.Left, std::min(newX, nextWorkArea.Right - width));
        newY = std::max(nextWorkArea.Top, std::min(newY, nextWorkArea.Bottom - height));

        win32.SetWindowRect(hwnd, newX, newY, width, height);
        Logger::Instance().Info(L"WindowManager", L"窗口已移动到下一个显示器");
    }

    void WindowManager::ExecutePreviousDisplay(int64_t targetHwnd)
    {
        int64_t hwnd = targetHwnd ? targetHwnd : GetTargetWindow();
        if (hwnd == 0) return;

        Win32WindowService win32;
        auto workAreas = win32.GetMonitorWorkAreas();
        if (workAreas.size() <= 1) return;

        auto currentWorkArea = GetTargetWorkArea(hwnd);
        int32_t currentIndex = -1;

        for (size_t i = 0; i < workAreas.size(); ++i)
        {
            if (workAreas[i].Left == currentWorkArea.Left &&
                workAreas[i].Top == currentWorkArea.Top)
            {
                currentIndex = static_cast<int32_t>(i);
                break;
            }
        }

        int32_t prevIndex = currentIndex <= 0 ? static_cast<int32_t>(workAreas.size()) - 1 : currentIndex - 1;
        auto prevWorkArea = workAreas[prevIndex];

        int32_t x, y, width, height;
        if (!win32.GetWindowRect(hwnd, x, y, width, height)) return;

        int32_t newX = prevWorkArea.Left + (x - currentWorkArea.Left);
        int32_t newY = prevWorkArea.Top + (y - currentWorkArea.Top);

        newX = std::max(prevWorkArea.Left, std::min(newX, prevWorkArea.Right - width));
        newY = std::max(prevWorkArea.Top, std::min(newY, prevWorkArea.Bottom - height));

        win32.SetWindowRect(hwnd, newX, newY, width, height);
        Logger::Instance().Info(L"WindowManager", L"窗口已移动到上一个显示器");
    }

    void WindowManager::ExecuteUndo()
    {
        if (m_operationHistory)
        {
            auto* historyManager = static_cast<OperationHistoryManager*>(m_operationHistory);
            if (historyManager->CanUndo())
            {
                historyManager->Undo();
                Logger::Instance().Info(L"WindowManager", L"撤销上一个操作");
                return;
            }
        }
        Logger::Instance().Info(L"WindowManager", L"没有可撤销的操作");
    }

    void WindowManager::ExecuteRedo()
    {
        if (m_operationHistory)
        {
            auto* historyManager = static_cast<OperationHistoryManager*>(m_operationHistory);
            if (historyManager->CanRedo())
            {
                historyManager->Redo();
                Logger::Instance().Info(L"WindowManager", L"重做上一个操作");
                return;
            }
        }
        Logger::Instance().Info(L"WindowManager", L"没有可重做的操作");
    }

    int64_t WindowManager::GetTargetWindow() const
    {
        if (m_cachedTargetWindow != 0 && Win32WindowService().IsWindow(m_cachedTargetWindow))
        {
            return m_cachedTargetWindow;
        }

        return GetTargetWindowCore();
    }

    int64_t WindowManager::GetTargetWindowCore() const
    {
        if (m_lastActiveService)
        {
            auto* lastActive = static_cast<LastActiveWindowService*>(m_lastActiveService);
            int64_t hwnd = lastActive->GetTargetWindow();
            if (hwnd != 0)
            {
                return hwnd;
            }
        }

        return Win32WindowService().GetForegroundWindow();
    }

    Core::WorkArea WindowManager::GetTargetWorkArea(int64_t hwnd) const
    {
        Win32WindowService win32;
        bool useCursorScreen = true;
        if (m_configService)
        {
            auto config = m_configService->Load();
            useCursorScreen = config.UseCursorScreenDetection;
        }

        return useCursorScreen
            ? win32.GetMonitorWorkAreaFromCursor()
            : win32.GetMonitorWorkAreaFromWindow(hwnd);
    }

    Core::WorkArea WindowManager::GetWorkAreaByDisplayIndex(int32_t index) const
    {
        Win32WindowService win32;
        auto workAreas = win32.GetMonitorWorkAreas();
        if (index < 0 || index >= static_cast<int32_t>(workAreas.size()))
        {
            return workAreas.size() > 0 ? workAreas[0] : Core::WorkArea(0, 0, 1920, 1080);
        }
        return workAreas[index];
    }

    std::pair<WindowAction, std::optional<int32_t>> WindowManager::GetActualAction(
        int64_t hwnd, WindowAction requestedAction, bool windowMovedExternally)
    {
        if (windowMovedExternally)
        {
            return { requestedAction, {} };
        }

        if (!m_configService)
        {
            return { requestedAction, {} };
        }

        auto config = m_configService->Load();
        auto mode = static_cast<Core::SubsequentExecutionMode>(config.SubsequentExecutionMode);

        if (mode != Core::SubsequentExecutionMode::CycleSize ||
            !Core::RepeatedExecutionsCalculator::SupportsCycle(requestedAction))
        {
            return { requestedAction, {} };
        }

        WindowHistory history;
        Core::WindowHistoryRecord record;
        if (!history.TryGetLastAction(hwnd, record))
        {
            return { requestedAction, {} };
        }

        if (record.Action != requestedAction)
        {
            return { requestedAction, {} };
        }

        Win32WindowService win32;
        auto workAreas = win32.GetMonitorWorkAreas();
        int32_t numDisplays = static_cast<int32_t>(workAreas.size());
        int32_t executionCount = record.Count + 1;

        if (numDisplays > 1 && Core::RepeatedExecutionsCalculator::SupportsCycle(requestedAction))
        {
            auto cycle = Core::RepeatedExecutionsCalculator::GetCycleGroup(requestedAction);
            int32_t groupLength = static_cast<int32_t>(cycle.size());
            int32_t totalCycleLength = groupLength * numDisplays;
            int32_t cycleIndex = (executionCount - 1) % totalCycleLength;
            int32_t displayIndex = cycleIndex / groupLength;
            int32_t positionInGroup = cycleIndex % groupLength;
            auto actualAction = cycle[positionInGroup];
            return { actualAction, displayIndex };
        }

        auto nextAction = Core::RepeatedExecutionsCalculator::GetNextCycleAction(requestedAction, executionCount);
        return { nextAction, {} };
    }

    Core::WindowRect WindowManager::ApplyWindowGap(Core::WindowRect target, const Core::WorkArea& workArea, WindowAction action)
    {
        if (m_gapSize <= 0) return target;

        const int32_t halfGap = m_gapSize / 2;

        switch (action)
        {
        case WindowAction::LeftHalf:
            target.Width -= halfGap;
            break;
        case WindowAction::RightHalf:
            target.X += halfGap;
            target.Width -= halfGap;
            break;
        case WindowAction::TopHalf:
            target.Height -= halfGap;
            break;
        case WindowAction::BottomHalf:
            target.Y += halfGap;
            target.Height -= halfGap;
            break;
        case WindowAction::TopLeft:
            target.Width -= halfGap;
            target.Height -= halfGap;
            break;
        case WindowAction::TopRight:
            target.X += halfGap;
            target.Width -= halfGap;
            target.Height -= halfGap;
            break;
        case WindowAction::BottomLeft:
            target.Y += halfGap;
            target.Width -= halfGap;
            target.Height -= halfGap;
            break;
        case WindowAction::BottomRight:
            target.X += halfGap;
            target.Y += halfGap;
            target.Width -= halfGap;
            target.Height -= halfGap;
            break;
        case WindowAction::FirstThird:
            target.Width -= halfGap;
            break;
        case WindowAction::CenterThird:
            target.X += halfGap;
            target.Width -= m_gapSize;
            break;
        case WindowAction::LastThird:
            target.X += halfGap;
            target.Width -= halfGap;
            break;
        default:
            break;
        }

        return target;
    }

    Core::WindowRect WindowManager::HandleFixedSizeWindow(const Core::WindowRect& current, const Core::WindowRect& target,
        const Core::WorkArea& workArea, WindowAction action)
    {
        switch (action)
        {
        case WindowAction::LeftHalf:
        case WindowAction::FirstThird:
        case WindowAction::FirstFourth:
            return Core::WindowRect(workArea.Left, target.Y, current.Width, current.Height);
        case WindowAction::RightHalf:
        case WindowAction::LastThird:
        case WindowAction::LastFourth:
            return Core::WindowRect(workArea.Right - current.Width, target.Y, current.Width, current.Height);
        case WindowAction::Center:
        case WindowAction::CenterHalf:
        case WindowAction::CenterThird:
        {
            int32_t centerX = workArea.Left + (workArea.Width() - current.Width) / 2;
            int32_t centerY = workArea.Top + (workArea.Height() - current.Height) / 2;
            return Core::WindowRect(centerX, centerY, current.Width, current.Height);
        }
        case WindowAction::TopHalf:
            return Core::WindowRect(target.X, workArea.Top, current.Width, current.Height);
        case WindowAction::BottomHalf:
            return Core::WindowRect(target.X, workArea.Bottom - current.Height, current.Width, current.Height);
        case WindowAction::MoveLeft:
        case WindowAction::MoveRight:
        case WindowAction::MoveUp:
        case WindowAction::MoveDown:
            return Core::WindowRect(target.X, target.Y, current.Width, current.Height);
        default:
        {
            int32_t centerX = workArea.Left + (workArea.Width() - current.Width) / 2;
            int32_t centerY = workArea.Top + (workArea.Height() - current.Height) / 2;
            return Core::WindowRect(centerX, centerY, current.Width, current.Height);
        }
        }
    }

    Core::WindowRect WindowManager::ApplyMinimumSize(Core::WindowRect target)
    {
        if (!m_configService) return target;

        auto config = m_configService->Load();
        int32_t minWidth = static_cast<int32_t>(config.MinimumWindowWidth);
        int32_t minHeight = static_cast<int32_t>(config.MinimumWindowHeight);

        if (target.Width < minWidth)
        {
            target.Width = minWidth;
        }
        if (target.Height < minHeight)
        {
            target.Height = minHeight;
        }

        return target;
    }

    Core::WindowRect WindowManager::ClampToWorkArea(Core::WindowRect target, const Core::WorkArea& workArea)
    {
        if (target.X < workArea.Left)
        {
            target.X = workArea.Left;
        }
        if (target.Y < workArea.Top)
        {
            target.Y = workArea.Top;
        }
        if (target.Right() > workArea.Right)
        {
            target.X = workArea.Right - target.Width;
        }
        if (target.Bottom() > workArea.Bottom)
        {
            target.Y = workArea.Bottom - target.Height;
        }

        return target;
    }

    void WindowManager::MoveCursorIfEnabled(int64_t hwnd, WindowAction action)
    {
        if (!m_configService) return;

        auto config = m_configService->Load();
        bool shouldMoveCursor = config.MoveCursor;

        if ((action == WindowAction::NextDisplay || action == WindowAction::PreviousDisplay)
            && !config.MoveCursorAcrossDisplays)
        {
            shouldMoveCursor = false;
        }

        if (shouldMoveCursor)
        {
            MoveCursorToWindowCenter(hwnd);
            Logger::Instance().Debug(L"WindowManager", L"光标已移动到窗口中心");
        }
    }

    void WindowManager::MoveCursorToWindowCenter(int64_t hwnd)
    {
        Win32WindowService win32;
        int32_t x, y, width, height;
        if (!win32.GetWindowRect(hwnd, x, y, width, height)) return;

        int32_t centerX = x + width / 2;
        int32_t centerY = y + height / 2;
        SetCursorPos(centerX, centerY);
    }

    void WindowManager::PlayBeep()
    {
        Beep(1000, 50);
    }

    void WindowManager::RecordToOperationHistory(WindowAction action, int64_t hwnd,
        const Core::WindowRect& oldRect, const Core::WindowRect& newRect,
        const std::wstring& processName)
    {
        if (!m_configService || !m_operationHistory)
        {
            return;
        }

        auto config = m_configService->Load();
        if (!config.History.Enabled)
        {
            return;
        }

        auto* historyManager = static_cast<OperationHistoryManager*>(m_operationHistory);
        historyManager->SetMaxHistoryCount(config.History.MaxHistoryCount);
        historyManager->RecordOperation(action, hwnd, oldRect, newRect, processName);
    }
}
